### PR TITLE
db cache

### PR DESCRIPTION
**Pull request by ilovers**

### CODE REVIEW

File: src/host/kvpair.rs [hunk 1]
```diff
@@ -88,24 +71,31 @@ impl<const DEPTH: usize> MongoMerkle<DEPTH> {
     ) -> Result<Option<MerkleRecord>, mongodb::error::Error> {
         let dbname = Self::get_db_name();
         let cname = self.get_collection_name();
-        let collection = get_collection::<MerkleRecord>(&self.client, dbname, cname)?;
-        let mut filter = doc! {};
-        filter.insert("index", index);
-        filter.insert("hash", bytes_to_bson(hash));
-        collection.find_one(filter, None)
+        let cache_key = cname.clone() + &index.to_string() + &hex::encode(hash);
```
Comment (yymone) [inline src/host/kvpair.rs:74]: This composing key logic is common for get and update record so can be a function.

File: src/host/kvpair.rs [hunk 2]
```diff
@@ -88,24 +71,31 @@ impl<const DEPTH: usize> MongoMerkle<DEPTH> {
...
+        let mut cache = MERKLE_CACHE.lock().unwrap();
+        if let Some(record) = cache.get(&cache_key) {
+            Ok(Some(record.clone()))
+        } else {
+            let collection = db::get_collection::<MerkleRecord>(dbname, cname)?;
+            let mut filter = doc! {};
+            filter.insert("index", index);
+            filter.insert("hash", bytes_to_bson(hash));
+            collection.find_one(filter, None)
```
Comment (yymone) [inline src/host/kvpair.rs:83]: Cache miss, so we load it from db, then do we need insert it into cache?

File: src/host/cache.rs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+use crate::host::kvpair::MerkleRecord;
+use lru::LruCache;
+use std::num::NonZeroUsize;
+use std::sync::Mutex;
+
+const DEFAULT_CACHE_SIZE: usize = 100000;
```
Comment (yymone) [inline src/host/cache.rs:6]: Some docs said:
"If maxsize is set to None, the LRU feature is disabled and the cache can grow without bound. The LRU feature performs best when maxsize is a power-of-two."
so maybe change to  2**17?

File: src/host/kvpair.rs (hunk 2 above)
Comment (ilovers) [reply src/host/kvpair.rs]: Yes, you are right. I have fixed it.

**Review by yymone — APPROVED** [review]

**Comment by xgaozoyoe** [comment]
Nice & merged.

**Pull request merged**